src/Context/DataProvider.js

```diff
@@ -0,0 +1,216 @@
+        };
+      case Actions.DECREASE_ITEM_IN_WISH:
+        return {
+          ...state,
+          wishProducts: removeOneItem(action.payload, state.wishProducts),
+        };
+      case Actions.INCREASE_ITEM_IN_CART:
+        return {
+          ...state,
+          cartProducts: increaseItem(action.payload, state.cartProducts),
+        };
+      case Actions.DECREASE_ITEM_IN_CART:
+        return {
+          ...state,
+          cartProducts: removeOneItem(action.payload, state.cartProducts),
+        };
+      default:
+        return state;
+    }
+  };
+
+  const [state, dispatch] = useReducer(dispatchFunction, {
+    products,
+    wishProducts: [],
+    cartProducts: [],
+  });
+
+  function addToCart(idFromCartClick, productItems, cartItems) {
+    const productIndex = productItems.findIndex(
+      (product) => product.id === idFromCartClick
+    );
+    const foundInCart = cartItems.findIndex(
+      (product) => product.id === idFromCartClick
+    );
+    if (foundInCart === -1) {
+      return [...cartItems, { ...productItems[productIndex], count: 1 }];
+    }
+    let allItems = cartItems.map((item) =>
+      item.id === idFromCartClick
+        ? { ...item, count: item.count + 1 }
+        : { ...item }
+    );
+    return allItems.filter((item) => item.count > 0);
+  }
+
+  function addToWish(idFromWishClick, wishItems, productItems) {
+    const productIndex = productItems.findIndex(
+      (product) => product.id === idFromWishClick
+    );
+    const foundInWish = wishItems.findIndex(
+      (product) => product.id === idFromWishClick
+    );
+    if (foundInWish === -1) {
+      return [...wishItems, { ...productItems[productIndex], count: 1 }];
+    }
+    let allItems = [...wishItems];
+    const index = allItems.findIndex((item) => item.id === idFromWishClick);
+    allItems[index].count += 1;
+    return allItems.filter((item) => item.count > 0);
+  }
+  function removeFromWish(idFromRemoveClick, wishItems) {
+    let allItems = [...wishItems];
+    const index = allItems.findIndex((item) => item.id === idFromRemoveClick);
+    allItems[index].count = 0;
+
+    return allItems.filter((item) => item.count > 0);
+  }
+
+  function moveToCart(idFromWish, countFromWish = 1, products, cartItems) {
+    const productIndex = products.findIndex(
+      (product) => product.id === idFromWish
+    );
+    const foundInCart = cartItems.findIndex(
+      (product) => product.id === idFromWish
+    );
+    if (foundInCart === -1) {
+      return [
+        ...cartItems,
+        {
+          ...products[productIndex],
+          count: countFromWish,
+        },
+      ];
+    }
+    const allItems = cartItems.map((item) =>
+      item.id === idFromWish
+        ? { ...item, count: item.count + countFromWish }
+        : { ...item }
+    );
+    return allItems.filter((item) => item.count > 0);
+  }
+  function moveToWish(idFromCart, countFromCart = 1, products, wishItems) {
```

Comment:
when you are having multiple parameters, it is a good thing to wrap them in object literal, then you don't have to worry abt the sequence
e.g. function moveToWish( {idFromCart, countFromCart, products, wishItems} )
And also if u have an object with the same property names u can pass as an argument by just spreading it
e.g. moveToWish({...product})

Reply:
Modified it 
Thanks for all the suggestions